source/vstd/arithmetic/power2.rs

```diff
@@ -239,4 +237,168 @@ pub proof fn lemma2_to64()
     ) by(compute_only);
 }
 
+/// Mask with low n bits set.
+pub open spec fn mask(n: nat) -> nat {
+    (pow2(n) - 1) as nat
+}
+
+/// Proof relating the n-bit mask to a function of the (n-1)-bit mask, for given
+/// n.
+pub proof fn lemma_mask_unfold(n: nat)
+    requires
+        n > 0,
+    ensures
+        mask(n) == 2 * mask((n - 1) as nat) + 1,
+{
+    calc!{ (==)
+        mask(n);
+            {}
+        (pow2(n) - 1) as nat;
+            { lemma_pow2_unfold(n); }
+        (2*pow2((n-1) as nat) - 1) as nat;
+            {}
+        (2*(pow2((n-1) as nat) - 1) + 1) as nat;
+            { lemma_pow2_pos((n-1) as nat); }
+        (2*mask((n-1) as nat) + 1) as nat;
+    }
+}
+
+/// Proof relating the n-bit mask to a function of the (n-1)-bit mask, for all
+/// n.
+pub proof fn lemma_mask_unfold_auto()
+    ensures
+        forall|n: nat| #![trigger mask(n)] n > 0 ==> mask(n) == 2 * mask((n - 1) as nat) + 1,
+{
+    assert forall|n: nat| n > 0 implies #[trigger] mask(n) == 2 * mask((n - 1) as nat) + 1 by {
+        lemma_mask_unfold(n);
+    }
+}
+
+/// Proof that mask(n) is odd, for given n.
+pub proof fn lemma_mask_is_odd(n: nat)
+    requires
+        n > 0,
+    ensures
+        mask(n) % 2 == 1,
+{
+    calc!{ (==)
+        mask(n) % 2;
+            { lemma_mask_unfold(n); }
+        (2 * mask((n-1) as nat) + 1) % 2;
+            { lemma_mod_multiples_vanish(mask((n-1) as nat) as int, 1, 2); }
+        1nat % 2;
+    }
+}
+
+/// Proof that mask(n) is odd, for all n.
+pub proof fn lemma_mask_is_odd_auto()
+    ensures
+        forall|n: nat| n > 0 ==> #[trigger] (mask(n) % 2) == 1,
+{
+    assert forall|n: nat| n > 0 implies #[trigger] (mask(n) % 2) == 1 by {
+        lemma_mask_is_odd(n);
+    }
+}
+
+/// Proof that for given n, dividing the low n bit mask by 2 gives the low n-1
+/// bit mask.
+pub proof fn lemma_mask_div2(n: nat)
+    requires
+        n > 0,
+    ensures
+        mask(n) / 2 == mask((n - 1) as nat),
+{
+    lemma_mask_unfold(n);
+}
+
+/// Proof that for all n, dividing the low n bit mask by 2 gives the low n-1
+/// bit mask.
+pub proof fn lemma_mask_div2_auto()
+    ensures
+        forall|n: nat| #![trigger mask(n)] n > 0 ==> mask(n) / 2 == mask((n - 1) as nat),
+{
+    assert forall|n: nat| n > 0 implies #[trigger] mask(n) / 2 == mask((n - 1) as nat) by {
+        lemma_mask_div2(n);
+    }
+}
+
+/// Proof establishing the concrete values of all masks of bit sizes from 0 to
+/// 32, and 64.
+pub proof fn lemma_mask_values()
```

Comment:
Here I copied the 0-32 and 64 range from `lemma2_to64()`.

I wonder if we should include all `0-64`? I also wondered if we could macro-generate a lemma like this with something like https://docs.rs/seq-macro/latest/seq_macro/, though this would require incurring a dependency or writing our own proc macro. (In practice I just wrote a quick Python script and copied it in.)

I'm also not sure about naming: `lemma2_to64` is an unusual name, and `lemma_mask_values` is maybe too vague.

Reply:
If you think people will want mask values in the 33-63 range, feel free to add the full set.  It does seem like something that should be done with a macro, however.  I'd suggest asking on #verus about taking a dependency on another crate.  For name, I'd suggested appending `_to64` to whatever name we end up with for `mask`.

Reply:
I tried to extend this to all values 0-64, however I appear to have triggered a bug in the process #1097 #1096. Perhaps leave this as-is for now?

Reply:
Okay, that's fine.